### ISSUE TITLE
invalid_arguments.cpp is busted

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#87514]
### 🐛 Describe the bug

This file is responsible for multi-overload error printing when arguments mismatch. It is supposed to have color formatting, where it tells you which argument mismatched, where green says it matched, and red says it mismatched. It is also completely broken.

<img width="946" alt="image" src="https://user-images.githubusercontent.com/13564/197295055-befa59a9-4de3-4dd3-bd95-5a3bd91d4c6a.png">

The brokenness appears to stem from the fact that, to determine how to match up the arguments, invalid_arguments.cpp *re-parses* the string names that the python arg parser assigns:

```
std::string FunctionParameter::type_name() const {
  switch (type_) {
    case ParameterType::TENSOR:
      return "Tensor";
    case ParameterType::SCALAR:
      return "Number";
    case ParameterType::INT64:
      return "int";
    case ParameterType::SYM_INT:
      return "int (or SymInt)";
    case ParameterType::DOUBLE:
      return "float";
    case ParameterType::COMPLEX:
      return "complex";
    case ParameterType::TENSOR_LIST:
      return "tuple of Tensors";
```

however, this parsing is completely busted; for example, to parse a tuple appropriately, you should must return a string `tuple[int]`. Furthermore, the names must exactly match what Python typename reports for the type in question.

This is completely bonkers because we have a type system for the parameters already in python_arg_parser.cpp, invalid_arguments.cpp should have reused it.

This would be a good starter task for someone, I think.

### Versions

master

**Comment by JDRanpariya** [comment]
I guess issue can be closed now or am I missing something?

**Comment by albanD** [comment]
I don't think this was fixed? Just tried locally and tuple(int,) is still in red while it is actually valid

**Comment by JDRanpariya** [comment]
Well I would like to take up the issue, and create a PR. I'm new to this and bit confused, I have built torch from source. Do I have to run both Python and C++ tests before I make any changes?

for the task if I understand correctly, we have to reuse type system that exists for python_arg_parser.cpp for invalid_arguments.cpp.

**Comment by albanD** [comment]
Hey!
The guide with all details on contributing is here: https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions :) 

I would say that this PR is not the easiest as your first one ;) 
If you do want to dive into this, the original comment does mention the cpp file within the python_arg_parser code that handles invalid arguments.
The part of the count_nonzero code you care about is actually generated during compilation and looks like this in the generated/python_torch_functionsX.cpp:
```cpp
// count_nonzero
static PyObject * THPVariable_count_nonzero(PyObject* self_, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "count_nonzero(Tensor input, int64_t? dim=None)",
    "count_nonzero(Tensor input, IntArrayRef dim)",
  }, /*traceable=*/true);

  ParsedArgs<2> parsed_args;
  auto _r = parser.parse(nullptr, args, kwargs, parsed_args);
  if(_r.has_torch_function()) {
    return handle_torch_function(_r, nullptr, args, kwargs, THPVariableFunctionsModule, "torch");
  }
  switch (_r.idx) {
    case 0: {
      // aten::count_nonzero(Tensor self, int? dim=None) -> Tensor
      
      auto dispatch_count_nonzero = [](const at::Tensor & self, ::std::optional<int64_t> dim) -> at::Tensor {
        pybind11::gil_scoped_release no_gil;
        return self.count_nonzero(dim);
      };
      return wrap(dispatch_count_nonzero(_r.tensor(0), _r.toInt64Optional(1)));
    }
    case 1: {
      // aten::count_nonzero.dim_IntList(Tensor self, int[] dim) -> Tensor
      
      auto dispatch_count_nonzero = [](const at::Tensor & self, at::IntArrayRef dim) -> at::Tensor {
        pybind11::gil_scoped_release no_gil;
        return self.count_nonzero(dim);
      };
      return wrap(dispatch_count_nonzero(_r.tensor(0), _r.intlist(1)));
    }
  }
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}
```

In particular here you see the overload definition and how the argparser is used. And where it fails to report the appropriate wrong types.
